src/OpenTelemetry/Batch.cs

```diff
@@ -140,6 +145,29 @@ public struct Enumerator : IEnumerator<T>
                 return false;
             };
 
+            private static readonly BatchEnumeratorMoveNextFunc MoveNextCircularBufferLogRecord = (ref Enumerator enumerator) =>
+            {
+                var circularBuffer = enumerator.circularBuffer;
+
+                var currentItem = enumerator.Current;
+                if (currentItem != null)
+                {
+                    if (typeof(T) == typeof(LogRecord))
+                    {
+                        LogRecordSharedPool.Current.Return((LogRecord)(object)currentItem);
```

Comment:
This extra casting bugs me. Likely not an easy fix for this.
I wonder if we would benefit from having a `class LogRecordBatch : Batch<LogRecord>` to avoid the extra type checks.
Probably out of scope for this PR. :)